support/BRFileService.c

```diff
@@ -127,34 +228,158 @@ fileServiceCreate (const char *basePath,
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_DELETE_ENTITY, -1, &fs->sdbDeleteStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_DELETE_ALL_TYPE_ENTITY, -1, &fs->sdbDeleteAllTypeStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    pthread_mutex_unlock (&fs->lock);
+
+    // Allocate the `entityTypes` array
+    array_new (fs->entityTypes, FILE_SERVICE_INITIAL_TYPE_COUNT);
+
+#if defined(DEBUG)
+    if (needSQLiteCompileOptions) {
+        needSQLiteCompileOptions = 0;
+        printf ("SQLITE ThreadSafe Mutex: %d\n", sqlite3_threadsafe());
+        printf ("SQLITE Compile Options:\n");
+        const char *option = NULL;
+        for (int index = 0;
+             NULL != (option = sqlite3_compileoption_get(index));
+             index++) {
+            printf ("-DSQLITE_%s\n", option);
+        }
+    }
+#endif
     return fs;
 }
 
+// This is callable with a partially allocated BRFileService.  So, be
+// careful with fields that might not yet exist.
 extern void
 fileServiceRelease (BRFileService fs) {
-    size_t typesCount = array_count(fs->entityTypes);
-    for (size_t index = 0; index < typesCount; index++)
-        fileServiceEntityTypeRelease (&fs->entityTypes[index]);
+    pthread_mutex_lock (&fs->lock);
+    if (NULL != fs->entityTypes) {
+        size_t typesCount = array_count(fs->entityTypes);
+        for (size_t index = 0; index < typesCount; index++)
+            fileServiceEntityTypeRelease (&fs->entityTypes[index]);
+        array_free(fs->entityTypes);
+    }
+
+    if (NULL != fs->network)  free (fs->network);
+    if (NULL != fs->currency) free (fs->currency);
+    if (NULL != fs->sdbPath)  free (fs->sdbPath);
+
+    if (NULL != fs->sdbInsertStmt)        sqlite3_finalize (fs->sdbInsertStmt);
+    if (NULL != fs->sdbSelectStmt)        sqlite3_finalize (fs->sdbSelectStmt);
+    if (NULL != fs->sdbSelectAllStmt)     sqlite3_finalize (fs->sdbSelectAllStmt);
+    if (NULL != fs->sdbUpdateStmt)        sqlite3_finalize (fs->sdbUpdateStmt);
+    if (NULL != fs->sdbDeleteStmt)        sqlite3_finalize (fs->sdbDeleteStmt);
+    if (NULL != fs->sdbDeleteAllTypeStmt) sqlite3_finalize (fs->sdbDeleteAllTypeStmt);
+    if (NULL != fs->sdbDeleteAllStmt)     sqlite3_finalize (fs->sdbDeleteAllStmt);
+
+    if (NULL != fs->sdb) sqlite3_close (fs->sdb);
+    pthread_mutex_unlock (&fs->lock);
+    pthread_mutex_destroy(&fs->lock);
 
-    free ((char *) fs->pathToType);
-    if (NULL != fs->entityTypes) array_free (fs->entityTypes);
     free (fs);
```

Comment:
What do you think about, as a pattern, `memset`-ing to 0 in release functions? Allows us to more easily catch use after frees.

Reply:
CORE-556